Implemented Search View

lucas-neuhaus-dev opened the pull request:
@marcaaron @shawnborton 

### Details
Replaced `ChatSwitcherView` with a newly implemented `SearchView` component. 
It shows a list with recent reports, with or without any messages sent.
If a report is selected, the search closes and we get redirected to it.

Deleted everything related to `ChatSwitcherView`, that would not be used anymore.

### Fixed Issues
Fixes #1338 

### Tests
1. Click the search icon;
2. Verify the list appear, with recent reports that contain or not a message sent;
3. Verify that after pressing / clicking in one of the rows, the report chat comes up.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots

#### Web
![web](https://i.ibb.co/bBfQ5FB/web.gif)

#### Mobile Web
![mobile web](https://i.ibb.co/tDYjnXW/mobile-web.gif)

#### Desktop
![desktop](https://i.ibb.co/6HHTZkp/desktop.gif)

#### iOS
![ios](https://i.ibb.co/84cFRcJ/ios.png)

#### Android
![android](https://i.ibb.co/dKRMsZp/android.gif)

lucas-neuhaus-dev committed:
Implemented Search View

shawnborton commented:
Any reason why my first and last name wouldn't be showing in the search rows?

lucas-neuhaus-dev commented:
What on earth, how did I not see that. Pause the review, I'm going to check it out, sorry.

lucas-neuhaus-dev commented:
@marcaaron 
About the Unit test error, I believe the test would actually need to be rewritten now, to get the desired output.
It is testing just with reports with no messages.

I'm going to take a look at that.

marcaaron commented on src/components/SearchView.js:
```diff
@@ -0,0 +1,152 @@
+     */
+    selectReport(option) {
+        redirect(ROUTES.getReportRoute(option.reportID));
+        this.reset();
```
Seems like this is the only place we call `reset()` should we just move this method into `selectReport()` and redirect once we have set `searchValue` back to `''`.

lucas-neuhaus-dev replied:
Done!

lucas-neuhaus-dev committed:
Changed getSearchOptions test to fit new desired output

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -158,6 +158,7 @@ function getOptions(reports, personalDetails, draftComments, activeReportID, {
     sortByLastMessageTimestamp = false,
     searchValue = '',
     showChatPreviewLine = false,
+    showEmptyRecentReports = false,
```
"empty" is maybe a bit ambiguous. Perhaps better to make this `showReportsWithNoComments` so that it is more clear when we check this:

```js
showReportsWithNoComments && hasNoComments && ...
```

lucas-neuhaus-dev replied:
That's definetly clearer. 
Done!

marcaaron commented on src/pages/NewChatPage.js:
```diff
@@ -73,8 +74,7 @@ class NewChatPage extends Component {
                 headerTitle: '',
-                // eslint-disable-next-line max-len
-                headerMessage: 'Don\'t see who you\'re looking for? Type their email or phone number to invite them to chat.',
+                headerMessage: CONST.MESSAGES.NO_CONTACTS_FOUND,
```
❤️

lucas-neuhaus-dev replied:
Promised you I wouldn't forget, haha

marcaaron reviewed (changes requested):
Looks great and I agree about the test. Just a small comment here and there.

lucas-neuhaus-dev commented:
@shawnborton 
The names are showing correctly. Updated images and gifs.

@marcaaron 
Updated `getSearchOptions` test to expect all reports, including the one with no message in it.

marcaaron commented on src/components/SearchView.js:
```diff
@@ -0,0 +1,152 @@
+    }
+}
+
+SearchView.propTypes = propTypes;
```
Adding this
```js
SearchView.displayName = 'SearchView';
``` 

will aid in debugging

lucas-neuhaus-dev replied:
Done!

lucas-neuhaus-dev committed:
Fixed reviews on SearchView

lucas-neuhaus-dev commented:
Reviews fixed.

marcaaron commented on src/components/SearchView.js:
```diff
@@ -0,0 +1,146 @@
+            title: 'RECENT',
+            data: this.state.recentReports,
+            shouldShow: this.state.recentReports.length > 0,
+            indexOffset: sections.reduce((prev, {data}) => prev + data.length, 0),
```
NAB, this would just be `0` in this case since there are no other sections.

lucas-neuhaus-dev replied:
:man_facepalming: <- me

marcaaron commented on src/components/SearchView.js:
```diff
@@ -0,0 +1,146 @@
+            indexOffset: sections.reduce((prev, {data}) => prev + data.length, 0),
+        });
+
+        return sections;
```
NAB, in fact, we can clean this up to just return the array with the one object in it I think.

lucas-neuhaus-dev replied:
Changed it to just return the simple array.

lucas-neuhaus-dev replied:
Left the title set in the object (not an empty string) just to make it very clear.

marcaaron reviewed (dismissed):
Everything looks good to me. Have some thoughts about future improvements that I will leave here. No need to tackle them in this PR.

- Searching is a bit slow when options are numerous. Probably we can fix this with some combination of debouncing and memoization.
- It takes some time to toggle the Search view (again noticeable with many options). I'm certain I caused this regression but the fix would be to stop conditionally rendering the the views in `SidebarLinks`
- We are discussing moving the `SearchPage` into a right docked modal for consistency. This would probably clean up the logic in the `HomePage` and the various checks for `ChatSwitcher.show()` etc
